apollo-federation/src/schema/validators/from_context.rs

```diff
@@ -137,17 +182,331 @@ static CONTEXT_PARSING_LEADING_PATTERN: LazyLock<Regex> =
+                                    context,
+                                    &frag_type_position,
+                                    &frag.selection_set,
+                                    schema,
+                                    target,
+                                    errors,
+                                ) {
+                                    // For inline fragments, remove NonNull wrapper as other subgraphs may not define this
+                                    // This matches the TypeScript behavior
+                                    if !expected_type.is_assignable_to(&resolved_type) {
+                                        errors.push(
+                                        SingleFederationError::ContextSelectionInvalid {
+                                            message: format!(
+                                                "Context \"{}\" is used in \"{}\" but the selection is invalid: the type of the selection \"{}\" does not match the expected type \"{}\"",
+                                                context, target, resolved_type, expected_type
+                                            ),
+                                        }
+                                        .into(),
+                                    );
+                                        return Ok(());
+                                    }
+                                } else {
+                                    errors.push(
+                                    SingleFederationError::ContextSelectionInvalid {
+                                        message: format!(
+                                            "Context \"{}\" is used in \"{}\" but the selection is invalid: the type of the selection does not match the expected type \"{}\"",
+                                            context, target, expected_type
+                                        ),
+                                    }
+                                    .into(),
+                                );
+                                    return Ok(());
+                                }
+                            }
+                        }
+                    }
+                    let context_location_names: std::collections::HashSet<String> =
+                        set_context_locations
+                            .iter()
+                            .map(|name| name.as_str().to_string())
+                            .collect();
+
+                    let mut has_matching_condition = false;
+                    for type_condition in type_conditions {
+                        if context_location_names.contains(type_condition) {
+                            has_matching_condition = true;
+                            break;
+                        }
+                    }
+
+                    if !has_matching_condition && !type_conditions.is_empty() {
```

Comment:
[JS code](https://github.com/apollographql/federation/blob/31104c6fc5319fb8140d697776249e9ee0ca9285/internals-js/src/federation.ts#L672) seems to check a different condition. It is checking if any of the inline spread type conditions (`type_conditions`) overlaps with the current location type `ty` (based on the possible runtime types check).